app-shell/src/notifications/store.ts

```
@@ -0,0 +1,299 @@
+        if (status === 'pending') {
+          pendingSubs.add(topic)
+        } else {
+          subscriptions.add(topic)
+          pendingSubs.delete(topic)
+        }
+        resolve()
+      } else {
+        reject(new Error('IP is not associated with a connection'))
+      }
+    })
+  }
+
+  public setUnubStatus(
+    ip: string,
+    topic: NotifyTopic,
+    status: 'pending' | 'unsubscribed'
+  ): Promise<void> {
+    return new Promise((resolve, reject) => {
+      if (ip in this.hosts) {
+        const { pendingUnsubs, subscriptions } = this.hosts[ip]
+        if (subscriptions.has(topic)) {
+          if (status === 'pending') {
+            pendingUnsubs.add(topic)
+          } else {
+            pendingUnsubs.delete(topic)
+            subscriptions.delete(topic)
+          }
+        }
+        resolve()
+      } else {
+        reject(new Error('IP is not associated with a connection'))
+      }
+    })
+  }
+
+  /**
+   *
+   * @description Creates a new hosts entry for a given IP with HostData that is a reference to an existing
+   * IP's HostData. This occurs when two IPs reported by discovery-client actually reference the same broker.
+   */
+  public associateIPWithExistingHostData(
+    ip: string,
+    robotName: string
+  ): Promise<void> {
+    return new Promise((resolve, reject) => {
+      const associatedHost = Object.values(this.hosts).find(
+        hostData => hostData.robotName === robotName
+      )
+      if (associatedHost != null) {
+        this.hosts[ip] = associatedHost
+        resolve()
+      } else {
+        reject(new Error('No associated IP found.'))
+      }
+    })
+  }
+
+  public deleteAllAssociatedIPsGivenRobotName(
+    robotName: string
+  ): Promise<void> {
+    return new Promise((resolve, reject) => {
+      const associatedHosts = this.getAssociatedIPsFromRobotName(robotName)
+      associatedHosts.forEach(hostname => {
+        delete this.hosts[hostname]
```

based on `associateIPWithExistingHostData`, won't it sometimes be the case that two ips refer to not just the same broker but the same js object by-ref? does deleting the same thing twice cause problems, or is `delete` here punned to just be "pop from object" or decref or something

---

Yeah, like a Set it just "fails silently", but weirdly enough it more accurately "succeeds silently", since deleting a non-existent property returns true when caste as a boolean. Fun stuff...